### PR TITLE
Channelwise Fully Connected Distconv Layer

### PR DESCRIPTION
**Pull request by szaman19**
- Adds distributed tensor implementation of channel-wise fully connected layer  
- Distributes input tensor in the channel dimension for a 4D input tensor of the form (Batch_size, channels, *, *)

### CODE REVIEW

File: include/lbann/layers/learning/distconv/distconv_layers.hpp [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+  get_fc_output_local_tensor_shape(const tensor::Tensor<DataType, locale, Allocator> &input,
+                                   const int_vector &linearity_dims,
+                                   bool transpose){
+
+    //https://github.com/LLNL/DiHydrogen/blob/7f86db1f9701ac3afb5e16aefdd57563d57a1698/legacy/include/distconv/distconv.hpp#L173
+
+    //Get the input layer local tensor shape 
+
+    auto output_local_shape = input.get_local_shape();
+    output_local_shape[0] = linearity_dims[0];
```
Comment (timmoon10) [inline include/lbann/layers/learning/distconv/distconv_layers.hpp:112]: I don't think we are handling the transpose case.

File: src/layers/learning/channelwise_fully_connected.cpp [hunk 2]
```diff
@@ -34,6 +34,297 @@
...
+  return data_type_distconv_adapter<TensorDataType>::get_activations_shape();
+}
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+dc::Shape
+channelwise_fully_connected_distconv_adapter<TensorDataType, Layout, Device>
+::get_prev_activations_shape(int index) const{
+  return data_type_distconv_adapter<TensorDataType>::get_prev_activations_shape(0);
+}
+
```
Comment (timmoon10) [inline src/layers/learning/channelwise_fully_connected.cpp:286]: If we are just calling the base class implementation, do we need to override these functions?
```suggestion
```

File: src/layers/learning/channelwise_fully_connected.cpp [hunk 3]
```diff
@@ -112,6 +403,51 @@ ::get_description() const
...
+::get_bias_dims() const
+{
+  const auto& output_dims = this->get_output_dims();
+  
+  const std::vector<size_t> output_channel_dims(
+    output_dims.begin()+1, output_dims.end());
+
+  const auto& output_channel_size = std::accumulate(
+    output_channel_dims.begin()+1, output_channel_dims.end(),
+    1, std::multiplies<size_t>());
```
Comment (timmoon10) [inline src/layers/learning/channelwise_fully_connected.cpp:444]:
```suggestion
  const auto& output_channel_size = std::accumulate(
    output_channel_dims.begin(), output_channel_dims.end(),
    1, std::multiplies<size_t>());
```

File: src/layers/learning/channelwise_fully_connected.cpp [hunk 4]
```diff
@@ -240,6 +589,20 @@ void
...
+#ifdef LBANN_HAS_DISTCONV
+  // We are guaranteed to have 
+  if(this->using_gpus() && this->distconv_enabled()){
+    this->get_distconv_adapter().fp_compute();
+    return ;
+  }else{
+    LBANN_ERROR("Distconv not compatible with CPU-only mode or distconv not enabled", 
+      "Is using GPU: \t ", this->using_gpus(),
+      "Is distconv enabled: \t" , this->distconv_enabled());
+  }
```
Comment (timmoon10) [inline src/layers/learning/channelwise_fully_connected.cpp:602]:
```suggestion
  if(this->distconv_enabled()){
    this->get_distconv_adapter().fp_compute();
    return ;
  }
```

File: src/layers/learning/channelwise_fully_connected.cpp [hunk 5]
```diff
@@ -315,6 +677,18 @@ void
...
 {
+#ifdef LBANN_HAS_DISTCONV
+
+  if(this->using_gpus() && this->distconv_enabled()){
+    this->get_distconv_adapter().bp_compute();
+
+    return ;
+  }else{
+    LBANN_ERROR("Distconv not compatible with CPU-only mode");
+  }
```
Comment (timmoon10) [inline src/layers/learning/channelwise_fully_connected.cpp:688]:
```suggestion
  if(this->distconv_enabled()){
    this->get_distconv_adapter().bp_compute();
    return ;
  }
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]

File: src/layers/learning/channelwise_fully_connected.cpp [hunk 6]
```diff
@@ -112,6 +403,51 @@ ::get_description() const
...
+  
+  const auto& input_channel_size = std::accumulate(
+    input_channel_dims.begin(), input_channel_dims.end(),
+    1, std::multiplies<size_t>());
+  
+  const auto& output_channel_size = std::accumulate(
+    output_channel_dims.begin(), output_channel_dims.end(),
+    1, std::multiplies<size_t>());
+  
+  std::vector<int> linearity_dims{1, 1, input_channel_size, output_channel_size};
```
[the remark this answers is not in the document]
Comment (szaman19) [reply src/layers/learning/channelwise_fully_connected.cpp:428]: We support it, but the linearity get's reshaped here: 

https://github.com/LLNL/lbann/blob/f5ec5481db47f1ef37208deceb56904ed32d6250/src/layers/learning/distconv/distconv_layers.cpp#L57

I think it makes sense to move that change dimensionality switch from the distconv side and put it in the `channelwise_fully_connected.cpp`.

File: src/layers/learning/distconv/distconv_layers.cpp [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+      util::MPIRootPrintStreamInfo()<< "output buffer is null";
+    }
+
+    if(!input.get_buffer() || input.get_buffer() == nullptr){
+      util::MPIRootPrintStreamInfo() << "input buffer is null";
+    }
+
+    if(!linearity.get_buffer() || linearity.get_buffer() == nullptr){
+      util::MPIRootPrintStreamInfo() <<"linearity buffer is null";
+    }
```
Comment (timmoon10) [inline src/layers/learning/distconv/distconv_layers.cpp:76]:
```suggestion
    if(output.get_buffer() == nullptr){
      util::MPIRootPrintStreamInfo()<< "output buffer is null";
    }

    if(input.get_buffer() == nullptr){
      util::MPIRootPrintStreamInfo() << "input buffer is null";
    }

    if(linearity.get_buffer() == nullptr){
      util::MPIRootPrintStreamInfo() <<"linearity buffer is null";
    }
```

File: src/layers/learning/distconv/distconv_layers.cpp [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+      local_mini_batch_size * num_local_channels,
+      output_grad_mat.LockedBuffer(),
+      output_size);
+    }
+    else {
+      El::Copy(output_grad_mat, output_grad_mat_reshaped);
+      output_grad_mat_reshaped.Resize(
+      output_size,
+      local_mini_batch_size * num_local_channels);
+    }
```
Comment (timmoon10) [inline src/layers/learning/distconv/distconv_layers.cpp:170]: This matrix doesn't seem to be used:
```suggestion
```

File: src/layers/learning/distconv/distconv_layers.cpp [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+    const auto& output_size = std::accumulate(output_dims.begin(), output_dims.begin()+1, 1, std::multiplies<size_t>());
+
+    const auto linearity_input_size = transpose_A ? output_size : input_size;
+    const auto linearity_output_size = transpose_A ? input_size : output_size;
+
+    const auto num_local_channels = output_dims[2];
+    const auto local_mini_batch_size = output_dims[3];
+
+    El::Matrix<DataType, El::Device::GPU>  input_mat_reshaped; 
+    El::Matrix<DataType, El::Device::GPU>  output_grad_mat_reshaped;
```
Comment (timmoon10) [inline src/layers/learning/distconv/distconv_layers.cpp:216]: These don't seem to be used:
```suggestion
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+    x1 = lbann.Reshape(lbann.Input(), dims=tools.str_list(_sample_dims), name="Input_layer")
+    x = lbann.Sum(x0, x1, name="Adding_weight_layer")
+    x_lbann = x
+
+    # Objects for LBANN model
+    obj = []
+    metrics = []
+    callbacks = []
+
+    num_height_groups = tools.gpus_per_node(lbann)
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:91]: Let's make this name more accurate:
```suggestion
    num_channel_groups = tools.gpus_per_node(lbann)
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+    x = (_samples
+         .reshape((-1, input_channel_size))
+         .transpose()
+         .astype(np.float64))
+    y = np.matmul(linearity.astype(np.float64), x)
+    z = tools.numpy_l2norm2(y) / _num_samples
+    val_without_bias = z
+
+    # ------------------------------------------
+    # Data-parallel distconv layout, non-transpose, bias
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:134]:
```suggestion
    # Non-transpose, bias
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+    y = lbann.ChannelwiseFullyConnected(
+        x,
+        weights=(linearity_weights, bias_weights),
+        output_channel_dims=output_channel_dims,
+        parallel_strategy=create_parallel_strategy(num_height_groups),
+        name="bias"
+    )
+    z = lbann.L2Norm2(y, name="L2Norm")
+    obj.append(z)
+    metrics.append(lbann.Metric(z, name='data-parallel layout, non-transpose, bias'))
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:160]:
```suggestion
    metrics.append(lbann.Metric(z, name='non-transpose, bias'))
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+    tol = 8 * val_with_bias * np.finfo(np.float32).eps
+    callbacks.append(lbann.CallbackCheckMetric(
+        metric=metrics[-1].name,
+        lower_bound=val_with_bias - tol,
+        upper_bound=val_with_bias + tol,
+        error_on_failure=True,
+        execution_modes='test'))
+
+    # ------------------------------------------
+    # Data-parallel distconv layout, non-transpose, no bias
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:172]:
```suggestion
    # Non-transpose, no bias
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+    y = lbann.ChannelwiseFullyConnected(
+        x,
+        weights=(linearity_weights),
+        output_channel_dims=output_channel_dims,
+        parallel_strategy=create_parallel_strategy(num_height_groups),
+        name="no_bias"
+    )
+    z = lbann.L2Norm2(y)
+    obj.append(z)
+    metrics.append(lbann.Metric(z, name='data-parallel layout, non-transpose, no bias'))
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:193]:
```suggestion
    metrics.append(lbann.Metric(z, name='non-transpose, no bias'))
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+    # NumPy implementation
+    tol = 8 * val_without_bias * np.finfo(np.float32).eps
+    callbacks.append(lbann.CallbackCheckMetric(
+        metric=metrics[-1].name,
+        lower_bound=val_without_bias - tol,
+        upper_bound=val_without_bias + tol,
+        error_on_failure=True,
+        execution_modes='test'))
+    # ------------------------------------------
+    # Data-parallel layout, transpose, bias
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:204]:
```suggestion
    # Transpose, bias
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+    y = lbann.ChannelwiseFullyConnected(
+        x,
+        weights=(linearity_weights, bias_weights),
+        output_channel_dims=output_channel_dims,
+        parallel_strategy=create_parallel_strategy(num_height_groups),
+        transpose=True,
+    )
+    z = lbann.L2Norm2(y)
+    obj.append(z)
+    metrics.append(lbann.Metric(z, name='data-parallel layout, transpose, bias'))
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:230]:
```suggestion
    metrics.append(lbann.Metric(z, name='transpose, bias'))
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+    tol = 8 * val_with_bias * np.finfo(np.float32).eps
+    callbacks.append(lbann.CallbackCheckMetric(
+        metric=metrics[-1].name,
+        lower_bound=val_with_bias-tol,
+        upper_bound=val_with_bias+tol,
+        error_on_failure=True,
+        execution_modes='test'))
+
+    # ------------------------------------------
+    # Data-parallel layout, transpose, no bias
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:242]:
```suggestion
    # Transpose, no bias
```

File: bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+        x,
+        weights=(linearity_weights),
+        output_channel_dims=output_channel_dims,
+        parallel_strategy=create_parallel_strategy(num_height_groups),
+        bias=False,
+        transpose=True
+    )
+    z = lbann.L2Norm2(y)
+    obj.append(z)
+    metrics.append(lbann.Metric(z, name='data-parallel layout, non-transpose, no bias'))
```
Comment (timmoon10) [inline bamboo/unit_tests/test_unit_layer_channelwise_fully_connected_distconv.py:263]:
```suggestion
    metrics.append(lbann.Metric(z, name='Non-transpose, no bias'))
```

**Review by timmoon10 — CHANGES_REQUESTED** [review]
One correctness bug and some style suggestions, but otherwise LGTM.

**Review by timmoon10 — APPROVED** [review]

**Pull request merged**